com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/graph/ResourceNode.java

```diff
@@ -0,0 +1,150 @@
+    private String absoluteFilepath;
+    private ResourceNode parent = null;
+    private IResource resource;
+    private String hexDigest = null;
+    private boolean excluded = false;
+    private final List<ResourceNode> children = new ArrayList<ResourceNode>();
+
+    public ResourceNode(IResource resource) {
+        this(resource.getPath(), resource.output().getAbsPath());
+        this.resource = resource;
+
+    }
+    public ResourceNode(final String relativeFilepath, final String absoluteFilepath) {
+        if (relativeFilepath.startsWith("/")) {
+            this.relativeFilepath = relativeFilepath;
+        } else {
+            this.relativeFilepath = "/" + relativeFilepath;
+        }
+        this.absoluteFilepath = absoluteFilepath;
+    }
+
+    public void addChild(ResourceNode childNode) {
+        childNode.parent = this;
+        this.children.add(childNode);
+    }
+
+    public void addChildren(List<ResourceNode> childNodes) {
+        for (ResourceNode childNode : childNodes) {
+            childNode.parent = this;
+            this.children.add(childNode);
+        }
+    }
+
+    public List<ResourceNode> getChildren() {
+        return this.children;
+    }
+
+    public ResourceNode getParent() {
+        return this.parent;
+    }
+
+    public IResource getResource() {
+        return this.resource;
+    }
+
+    public String getRelativePath() {
+        return relativeFilepath;
+    }
+
+    public String getAbsolutePath() {
+        return absoluteFilepath;
+    }
+
+    public void setHexDigest(String hexDigest) {
+        this.hexDigest = hexDigest;
+    }
+
+    public void setExcluded(boolean excluded) {
+        this.excluded = excluded;
+    }
+
+    private void writeJSON(JsonGenerator generator) throws IOException {
+        generator.writeStartObject();
+        generator.writeFieldName("path");
+        generator.writeString(relativeFilepath);
+
+        generator.writeFieldName("hexDigest");
+        generator.writeString(hexDigest);
+
+        generator.writeFieldName("excluded");
+        generator.writeBoolean(excluded);
+
+        generator.writeFieldName("children");
+        generator.writeStartArray();
+        for (ResourceNode child : children) {
+            child.writeJSON(generator);
+        }
+        generator.writeEndArray();
+        generator.writeEndObject();
+    }
+
+    public String toJSON() throws IOException {
+        StringWriter stringWriter = new StringWriter();
+        BufferedWriter writer = new BufferedWriter(stringWriter);
+        JsonGenerator generator = null;
+        try {
+            generator = (new JsonFactory()).createJsonGenerator(writer);
+            generator.useDefaultPrettyPrinter();
+            writeJSON(generator);
+        }
+        finally {
+            if (generator != null) {
+                generator.close();
+            }
+            IOUtils.closeQuietly(writer);
+        }
+        return stringWriter.toString();
+    }
```

Comment:
Do we want a function to write directly to a file, recursively?

Currently, we create a bigger, and bigger json string, for each node. What will the performance be like on a large project?

Reply:
I suggest testing/timing on Family Island to get a feeling for it.

You return "stringWriter.toString()", I assume it allocates a String, with the actual string, and the writer can be garbage collected?